src/serial_port.rs

```diff
@@ -93,6 +94,47 @@ impl SerialPort {
 		self.inner.is_read_vectored()
 	}
 
+	/// Will read a byte at a time until either a timeout is triggered, the size is reached, or memory is exhausted.
+	/// This differs from `read()` which will either timeout or read a single byte.
+	/// # Examples
+	///
+	/// ```
+	/// use SerialTester::read_to_timeout;
+	///
+	/// let mut port = serial2::Serialport::open("COM2", 9600);
+	/// assert_eq!(read_to_timeout(&mut port, None), );
+	/// assert_eq!(port, [0x48, 0x49, 0x20, 0x46, 0x45, 0x52, 0x52, 0x49, 0x53]);
+	/// ```
+	pub fn read_until_timeout(&self, size: Option<NonZeroUsize>) -> std::io::Result<Vec<u8>> {
```

Comment:
Instead of this 
```
pub fn read_until_timeout(&self, size: Option<NonZeroUsize>) -> std::io::Result<Vec<u8>> {
```

You could do this
```
pub fn read_until_timeout(&self, buffer: &mut [u8]) -> usize{
```

the function would then read into the buffer slice and return the number of bytes successfully into the buffer. If a timeout occured, then obviously the return value will < buff.len()

this aligns with how many other apis work as well, and is also nostd ( embedded ) friendly as the buffer could just be a slice into a static allocated array

Reply:
Also, the timeout bounds is kinda common for all of these methods, so does not really need to be explained.

pub fn read_into(...)

It also might make sense to instead impl the read trait:

https://doc.rust-lang.org/std/io/trait.Read.html

fn [read_exact](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_exact)(&mut self, buf: &mut [[u8](https://doc.rust-lang.org/std/primitive.u8.html)]) -> [Result](https://doc.rust-lang.org/std/io/type.Result.html)<[()](https://doc.rust-lang.org/std/primitive.unit.html)> { ... }

In such a case, ReadInterrupted would be returned for a read interrupted by Timeout

Reply:
My primary use is an API that doesn't really care about the size and more about the timeout, which is why I named it such.

You do showcase something interesting which is that this API is in essence very similar to read_exact. Just it special cases timeout as being Ok, which you can do manually with something like

```rust
match port.read_exact(&mut buf) {
    Ok(_) => (),
    Err(err) if err.kind() == std::io::ErrorKind::TimedOut => (),
    Err(err) => return Err(err),
}
```
Though this does rely on the implementation detail of this library which is that the buffer is written to in the loop.

Reply:
`read_exact` using a loop is something you can rely on. There is no other way, since the contract of the underlying `read` syscall is that it can return less than the requested bytes. At-least on Unix, and we try to get that behavior on Windows too.

I'm still considering if this should be fixed by a timeout returned `Ok(0)` instead of an error or not.

If we do, `read_exact` will still fail with `UnexpectedEof` (instead of `TimedOut`). But `read_to_end()` will return `Ok`.

From the standard library documentation of `read()`:
> If the return value of this method is [Ok(n)](https://doc.rust-lang.org/stable/std/result/enum.Result.html#variant.Ok), then implementations must guarantee that 0 <= n <= buf.len(). A nonzero n value indicates that the buffer buf has been filled in with n bytes of data from this source. If n is 0, then it can indicate one of two scenarios:
>
>    This reader has reached its “end of file” and will likely no longer be able to produce bytes. Note that this does not mean that the reader will always no longer be able to produce bytes. As an example, on Linux, this method will call the recv syscall for a [TcpStream](https://doc.rust-lang.org/stable/std/net/struct.TcpStream.html), where returning zero indicates the connection was shut down correctly. While for [File](https://doc.rust-lang.org/stable/std/fs/struct.File.html), it is possible to reach the end of file and get zero as result, but if more data is appended to the file, future calls to read will return more data.
    The buffer specified was 0 bytes in length.

So the question is: should we mimick the guarantees of a `TcpStream` or a `File`? Conceptually, a serial port is more similar to a `TcpStream` than a `File`, so that kinda hints to keeping the timeout as an error instead of `Ok(0)`.

But then a convenience function that reads until a timeout (and returns `Ok`) is certainly useful.